src/LanguageServer/Protocol/Handler/DocumentChanges/DidChangeHandler.cs

```diff
@@ -63,20 +70,41 @@ private static SourceText GetUpdatedSourceText(TextDocumentContentChangeEvent[]
         // If the host sends us changes in a way such that no earlier change can affect the position of a later change,
         // then we can merge the changes into a single TextChange, allowing creation of only a single new
         // source text.
-        if (AreChangesInReverseOrder(contentChanges))
+        if (AreChangesInReverseOrder(remainingContentChanges))
         {
             // The changes were in reverse document order, so we can merge them into a single operation on the source text.
             // Note that the WithChanges implementation works more efficiently with it's input in forward document order.
-            var newChanges = contentChanges.Reverse().SelectAsArray(change => ProtocolConversions.ContentChangeEventToTextChange(change, text));
+            var newChanges = remainingContentChanges.Reverse().SelectAsArray(change => ProtocolConversions.ContentChangeEventToTextChange(change, text));
             text = text.WithChanges(newChanges);
         }
         else
         {
             // The host didn't send us the items ordered, so we'll apply each one independently.
-            foreach (var change in contentChanges)
+            foreach (var change in remainingContentChanges)
                 text = text.WithChanges(ProtocolConversions.ContentChangeEventToTextChange(change, text));
         }
 
         return text;
     }
+
+    private static (ImmutableArray<TextDocumentContentChangeEvent>, SourceText) GetUpdatedSouorceTextAndChangesAfterFullTextReplacementHandled(SumType<TextDocumentContentChangeEvent, TextDocumentContentChangeFullReplacementEvent>[] contentChanges, SourceText text)
+    {
+        // Per the LSP spec, each content change can be either a TextDocumentContentChangeEvent or TextDocumentContentChangeFullReplacementEvent.
+        // The former is a range-based change while the latter is a full text replacement. If a TextDocumentContentChangeFullReplacementEvent is found,
+        // then make a full text replacement for that and return all subsequent changes as remaining range-based changes.
+        var lastFullTextChangeEventIndex = contentChanges.Length - 1;
+        for (; lastFullTextChangeEventIndex >= 0; lastFullTextChangeEventIndex--)
+        {
+            var change = contentChanges[lastFullTextChangeEventIndex];
+            if (change.Value is TextDocumentContentChangeFullReplacementEvent onlyTextEvent)
+            {
+                // Found a full text replacement. Create the new text and stop processing.
+                text = text.WithChanges([new TextChange(new TextSpan(0, text.Length), onlyTextEvent.Text)]);
```

Comment:
Why doing the WithChanges versus just creating a new text from scratch?

Reply:
I didn't want to have to duplicate the parameter knowledge invoked with doing that as done here:

https://github.com/dotnet/roslyn/blob/4f8f5e6feb80a9c2d651232735ef69d32710927f/src/LanguageServer/Protocol/Handler/DocumentChanges/DidOpenHandler.cs#L39